File: python/leetcode/problem39.py
def search(c, i, sum, t, cur_list, res_list):
    if sum > t:
        return
    if sum == t:
        res_list.append(cur_list.copy())
        return
    k = i
    while k < len(c):
        cur_list.append(c[k])
        search(c, k, sum + c[k], t, cur_list, res_list)
        cur_list.pop()
        k += 1

class Solution:
    def combinationSum(self, candidates, target):
        res_list = []
        search(candidates, 0, 0, target, [], res_list)
        return res_list

File: intelliJ/src/main/java/LeetCode/Problem39.java
package LeetCode;

import java.util.ArrayList;
import java.util.List;
import java.util.Vector;

public class Problem39 {
    private void search(int[] c, int i, int sum, int target, List<Integer> curList, List<List<Integer>> resList) {
        if (sum > target) {
            return;
        }

        if (sum == target) {
            resList.add(curList);
        }

        int k = i;
        while (k < c.length) {
            List<Integer> nextList = new Vector<>(curList);
            nextList.add(c[k]);
            search(c, k, sum + c[k], target, nextList, resList);
            k += 1;
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> resList = new Vector<>();
        search(candidates, 0, 0, target, new Vector<>(), resList);
        return resList;
    }
}
